### ISSUE TITLE
Implement Optional, Vector and Dictionary settings controls in Javascript

### ISSUE THREAD

**Issue by zarik5** [alvr-org/ALVR#266]
This is needed to expose custom controller mappings and FFmpeg encoder/decoder settings.

**Comment by JackD83** [comment]
Can you please provide the schema and a session containing those values? Implementing them should not be that hard

**Comment by zarik5** [comment]
You can add the following code for testing in a struct in `settings.rs`:
```rust
pub optional_test: Option<CodecType>,
pub vector_test: Vec<CodecType>,
pub dictionary_test: Vec<(String, CodecType)>,
```

In `settings_cache_default()` add the corresponding:
```rust
optional_test: OptionalDefault {
    set: true,
    content: CodecTypeDefault {
        variant: CodecTypeDefaultVariant::H264,
    },
},
vector_test: VectorDefault {
    element: CodecTypeDefault {
        variant: CodecTypeDefaultVariant::H264,
    },
    default: vec![CodecType::H264],
},
vector_test: DictionaryDefault {
    key: "Test".into(),
    value: CodecTypeDefault {
        variant: CodecTypeDefaultVariant::H264,
    },
    default: vec![("ABC".into(), CodecType::H264)],
},
```

The Optional should appear as a button group: ( set | unset )
The Vector should appear as a one column table where the last row has a (+) icon.
The Dictionary should appear as a two column table where the last row has a (+) icon.

**Comment by zarik5** [comment]
For the network rewrite (v13) I only need the Optional. We'll leave the vector and dictionary for the encoder rewrite.

**Comment by JackD83** [comment]
Whats the difference between "optional" and "switch"?

**Comment by zarik5** [comment]
Switch and Option are both enums with one variant with a value and one variant without value, so they are functionally equivalent but they have different meanings.
Switch is an enum that I defined. It should be used when the value (enabled or disabled) must always be specified.
Optional reuses the the built-in Rust Option and should be used when you either want to specify a value or you want to leave it unset (then the default value is used). Just to give you an example: using the async runtime tokio, I can either create the runtime without parameters, or I can specify the maximum number of CPU cores used. In the settings I can then define: `runtime_cores: Option<u8>`.

**Comment by JackD83** [comment]
That is the internal usage. I was just wondering if there is a difference in the UI between the two. If not, they could share the same code

**Comment by zarik5** [comment]
Switch is displayed as a checkbox. Optional should be displayed as a stateful button group: ( set | unset ). Also both the schema and the session is different. So I don't think you can reuse code.

**Comment by JackD83** [comment]
To b honest, I still have no idea how to implement this. I added the parameters to settings.rs and get the following schema:

```JavaScript
[
    "optionalTest",
    {
        "advanced": false,
        "content": {
            "type": "optional",
            "content": {
                "defaultSet": true,
                "content": {
                    "type": "choice",
                    "content": {
                        "default": "H264",
                        "variants": [
                            [
                                "H264",
                                null
                            ],
                            [
                                "HEVC",
                                null
                            ]
                        ]
                    }
                }
            }
        }
    }
]
```


That is a boolean for default set and a content, it that case a choice. This would result in something like this:

<img width="424" alt="optional1" src="https://user-images.githubusercontent.com/15889805/89759588-c9bc8880-daea-11ea-8485-44dbdfa2a131.PNG">

A button group is something like this:

<img width="295" alt="buttongrp" src="https://user-images.githubusercontent.com/15889805/89759709-11431480-daeb-11ea-89fb-1b0c216a6b06.PNG">

I don't see how the schema translates to a button group

**Comment by zarik5** [comment]
When the defaultset/set is true, highlight the button "set", when defaultset/set is false, highlight the button "unset".
Sorry if i'm a bit pedantic with the distinction between Optional and Switch, but I think that is for the benefit of the user and the clarity of the backend code.

**Comment by JackD83** [comment]
Now I understand :) No its not pedantic, it was just a little unclear what you envisioned. 
I think its still the same code as switch, but with a different representation of the checkbox.

Optional has two buttons in a button group which toggle each other, one saying "set" the other "unsest"
<img width="113" alt="optional2" src="https://user-images.githubusercontent.com/15889805/89775753-c1734600-db08-11ea-8721-ae06b1eeface.png">

**Comment by zarik5** [comment]
That's perfect! Thank you

**Comment by zarik5** [comment]
@JackD83 Without opening another issue, can you check if doing something like `Option<Switch<u64>>` works? Until now I've been careful to never nest two non Section schema elements. Doing something like `Option<Switch<u64>>` is not essential but not going through a section would make the code more concise.

**Comment by ronthecookie** [comment]
Merged into #674.
